plugins/ui/edgy.nix

```diff
@@ -0,0 +1,323 @@
+          Whether to enable animations.
+        '';
+
+        fps = helpers.defaultNullOpts.mkUnsignedInt 100 ''
+          Frames per second.
+        '';
+
+        cps = helpers.defaultNullOpts.mkUnsignedInt 120 ''
+          Cells per second.
+        '';
+
+        on_begin = helpers.defaultNullOpts.mkLuaFn ''
+          function()
+            vim.g.minianimate_disable = true
+          end
+        '' "Callback for the beginning of animations.";
+
+        on_end = helpers.defaultNullOpts.mkLuaFn ''
+          function()
+            vim.g.minianimate_disable = false
+          end
+        '' "Callback for the ending of animations.";
+
+        # This option accepts an attrs or a lua string.
+        # Hence, we use `mkOption` to convert the string to raw lua in `apply`.
+        spinner = mkOption {
+          type =
+            with helpers.nixvimTypes;
+            nullOr (
+              either strLua (submodule {
+                freeformType = attrsOf anything;
+                options = {
+                  frames = helpers.defaultNullOpts.mkListOf types.str [
+                    "⠋"
+                    "⠙"
+                    "⠹"
+                    "⠸"
+                    "⠼"
+                    "⠴"
+                    "⠦"
+                    "⠧"
+                    "⠇"
+                    "⠏"
+                  ] "Frame characters.";
+
+                  interval = helpers.defaultNullOpts.mkUnsignedInt 80 ''
+                    Interval time between two consecutive frames.
+                  '';
+                };
+              })
+            );
+          default = null;
+          example = "require('noice.util.spinners').spinners.circleFull";
+          apply = v: if isString v then helpers.mkRaw v else v;
+          description = helpers.defaultNullOpts.mkDesc {
+            frames = [
+              "⠋"
+              "⠙"
+              "⠹"
+              "⠸"
+              "⠼"
+              "⠴"
+              "⠦"
+              "⠧"
+              "⠇"
+              "⠏"
+            ];
+            interval = 80;
+          } "Spinner for pinned views that are loading.";
+        };
+      };
+
+      exit_when_last = helpers.defaultNullOpts.mkBool false ''
+        Enable this to exit Neovim when only edgy windows are left.
+      '';
+
+      close_when_all_hidden = helpers.defaultNullOpts.mkBool true ''
+        Close edgy when all windows are hidden instead of opening one of them.
+        Disable to always keep at least one edgy split visible in each open section.
+      '';
+
+      wo = helpers.defaultNullOpts.mkAttrsOf types.anything {
+        winbar = true;
+        winfixwidth = true;
+        winfixheight = false;
+        winhighlight = "WinBar:EdgyWinBar,Normal:EdgyNormal";
+        spell = false;
+        signcolumn = "no";
+      } "Global window options for edgebar windows.";
+
+      keys = mkOption {
+        type = with helpers.nixvimTypes; attrsOf (either strLuaFn (enum [ false ]));
```

Comment:
I'd be tempted to move `with helpers.nixvimTypes` so that it applies to the whole of `settingsOptions`.

Reply:
Specifically for this plugin or in general ?

Reply:
In general, if I found I was referencing types/nixvimTypes a fair amount throughout `settingsOptions`.

Reply:
> However, this practice (of relying on `with` statements) is being discouraged in nixpkgs as it can lead to subtle and nasty errors.

I agree with that principal, but I think using a `with types` statement on an attrset exclusively containing option definitions is a reasonable exception.

Reply:
As I say, just my preference. No strong opinions.